### ISSUE TITLE
Differentiate between scopes

### ISSUE THREAD

**Issue by zunder** [area/language-latex#124]
I have problems using language-tex in combination with [spell-check](https://github.com/atom/spell-check). Spell-check allows to define the scope it should make the suggestions for. Unfortunately language-tex provides only a `text.tex.latex` scope, which is causing spelling check in LaTeX commands as well as in regular text.

![2017-03-23-101817_737x264_scrot](https://cloud.githubusercontent.com/assets/6862750/24240508/303267e0-0fb2-11e7-9694-ba0ff05912a7.png)

Is there a way to make a differentiation of scope between commands and regular text?

**Comment by PierreMarchand20** [comment]
I am waiting for this to be included in atom to solve the problem you mentioned : https://github.com/atom/spell-check/pull/157

**Comment by Aerijo** [comment]
Solution: don't use spell-check. Use the linter, linter-ui-default, linter-spell, linter-spell-latex package combination, as the last one is specifically made for compatibility with this package. You do need hunspell or aspell on your system though.

If you must use spell-check, I'm not actually sure what the problem is: your screenshot shows the macro to be scoped as meta.function and support.function. Adding one of these to the exclusion list should work.

**Comment by PierreMarchand20** [comment]
We cannot add this scope to the exclusion list since there is no exclusion list at the moment as far as I know. There will be one when atom/spell-check#157 will be merged.

**Comment by Aerijo** [comment]
Fair enough. I was thinking of the autocomplete-plus package, where a simiar thing exists for snippets.

I still recommend using the other package combination though, as spell-check doesn't seem aimed at supporting latex. It does mention a latex use case, but using linter etc. also lets you use the chktex package, which is probably useful.

**Comment by yudai-nkt** [comment]
I don't think there is appropriate scopes to distinguish those two, but will consider some solutions after atom/spell-check#157 is merged. Please use other lint packages meanwhile.

By the way, I believe order of growth should be denoted in math mode i.e., `$O(n)$`...

**Comment by PierreMarchand20** [comment]
The feature described in atom/spell-check#157 will be in Atom 1.25. I still think it will be great to have a great compatibility with spell-check since it is a default package. (I must admit I tried the solution described with Aerijo with linter without success, and it seemed pretty heavy to handle compare to spell-check, even if, of course, it seems more powerful)

**Comment by Aerijo** [comment]
@PierreMarchand20 I'm not actually sure what is being asked for here. Every language has a 'root' scope that is applied to all the text. In this case, the root scope is `text.tex.latex`. A lack of additional scopes indicates the word is normal text, while scopes such as `support.function.textit.latex` indicate it is a control word.

There is no other way to differentiate text, so a scope blacklist is the only way. Adding a unique scope to all text that is not otherwise scoped would 1) be difficult to achieve 2) probably cause a lot of bugs, and 3) be made redundant when Atom 1.25 is released and `spell-check` gets a scope blacklist.

**Comment by Aerijo** [comment]
Also, I can try my best to help with any problems you have setting up `linter-spell-latex`. Was there a step in particular, or would a general guide be helpful?

**Comment by PierreMarchand20** [comment]
Concerning `spell-check`, I do not know what would be possible. I guess all the scopes (other than `text.tex.latex`) could be added to the scope blacklist by default in Atom 1.25 ? or at least have a list of all the scopes that we should add. I mean, it would be great to have 'language-latex' directly work nicely with default features in Atom.

Concerning 'linter-spell-latex', I installed everything you said and at the end I did not succeed to make it work. I do not remember exactly why but I will try again.

**Comment by Aerijo** [comment]
@PierreMarchand20 Maybe see if [this guide I wrote](https://gist.github.com/Aerijo/5b9522530715e5be6e89fc012e9a72a8#spelling) can help. If it doesn't, let me know (there or here) and I'll do my best to improve it.

**Comment by PierreMarchand20** [comment]
Thank you ! it is an amazing work, I did not know about hyperclick and autocomplete, I will check it out. Actually I was using `latextools`, but thanks to your guide, I noticed it was not maintained and now I use `latex`.

I tried `linter-spell-latex` but, it just does not work. If I put an error, it does not find it. I do not understand why. May be my project is to big and it has an hard time spell-checking all my files, I do not know... I will check if there is an issue about it in the `linter-spell-latex` repo.

**Comment by Aerijo** [comment]
@PierreMarchand20 My bad, I forgot an important package for the linting. You also need [`linter-iu-default`](https://atom.io/packages/linter-ui-default), which is the one that actually shows the spelling mistakes (in the user interface). I'll update the guide to reflect this tomorrow.

**Comment by PierreMarchand20** [comment]
Ok I succeeded in using `linter-spell-latex` but it was not so easy. There is a bug, I think, when using `aspell` with multiple dictionaries, then I used hunspell but I had to add the dictionaries myself (I use macOS and I want to be able to use fr_FR and en_US). So now it is working, but having `spell-check` working is still a good idea for those who do not want to spend a lot of time on this kind of thing :-) Thank you for your help !
